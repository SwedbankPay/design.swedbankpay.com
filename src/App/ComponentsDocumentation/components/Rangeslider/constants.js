import React from "react";
import Rangeslider from "@components/Rangeslider";

export const howToUse = [
    {
        type: "success",
        component: <Rangeslider label="Label" min={0} max={200} value={100}/>,
        text: <p>Do use <b>sliders</b> for content with numeric values.</p>
    },
    {
        type: "success",
        component: <Rangeslider label="Label" min={0} max={100} value={50} valueLabelPostfix="%"/>,
        text: <p>The <b>sliders</b> component also have a post label for special symbols to help the user easier understand the slider value. It could be shown as above with the percentage symbol.</p>
    },
    {
        type: "success",
        component: <Rangeslider label="Label" min={0} max={200} value={100} valueLabelPrefix="$" />,
        text: <p>It also have a pre label for other symbols such as currency e.g. €, $ and £.</p>
    },
    {
        type: "error",
        component: <Rangeslider label="Label" value={50} />,
        text: <p>Don’t modify and remove label and maximum/minimum indicators as this makes it harder for user to grasp the content and context of use.</p>
    },
    {
        type: "error",
        component:
            <div>
                <div className="rangeslider">
                    <input type="range"/>
                    <div className="d-flex justify-content-between">
                        <p className="rangeslider-label">Label</p>
                        <output className="value-label px-2 align-items-center">
                            Average
                        </output>
                    </div>
                    <div className="value-range">
                        <small>Poor</small>
                        <small>Great</small>
                    </div>
                </div>
            </div>,
        text: <p>Don’t use sliders for content without numeric values.</p>
    }
];

export const sliderShowCase = {
    id: "no-tabs",
    elements: [
        {
            component: <Rangeslider label="Label" min={0} max={100} value={50}/>,
            options: {
                checkbox: [
                    {
                        title: "Label add on",
                        inputs: [
                            {
                                id: "help_icon",
                                name: "With help icon",
                                value: {
                                    tooltip: true
                                }
                            }
                        ]
                    }
                ],
                radio: [
                    {
                        id: "unit_radio",
                        title: "Unit value",
                        values: [
                            {
                                name: "Default",
                                value: {
                                    valueLabelPrefix: "",
                                    valueLabelPostfix: ""
                                }
                            },
                            {
                                name: "Currency",
                                value: {
                                    valueLabelPrefix: "$"
                                }
                            },
                            {
                                name: "Percent",
                                value: {
                                    valueLabelPostfix: "%"
                                }
                            }
                        ]
                    }
                ]
            },
            title: "Slider",
            description: "The slider component has a label which decribes the intent of the slider. It also has a minimum and maximum range which displays the range which user can pick inbetween. By holding and draging the circle indicator along the line, the user can see the value change and update in real time."
        }
    ]
};
