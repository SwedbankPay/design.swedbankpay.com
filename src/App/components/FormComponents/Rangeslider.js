import React from "react";
import PropTypes from "prop-types";

const Rangeslider = ({ label, name, min, max, step, value, disabled, valueLabelPrefix, valueLabelPostfix, tooltip }) => {
    const attrs = {
        type: "range",
        name,
        min,
        max,
        step,
        defaultValue: value,
        disabled
    };

    const ValueLabel = () => (
        <output className="value-label" >{"\n\t\t"}
            <p>{"\n\t\t\t"}
                {valueLabelPrefix ? <span>{valueLabelPrefix}</span> : null}{valueLabelPrefix ? "\n\t\t\t" : null}
                <span data-rs-value>{value}</span>{"\n\t\t\t"}
                {valueLabelPostfix ? <span>{valueLabelPostfix}</span> : null}{valueLabelPostfix ? "\n\t\t" : null}
            </p>{"\n\t"}
        </output>
    );

    return (
        <div className={`rangeslider ${disabled ? "disabled" : ""}`}>{"\n"}
            <label htmlFor="rangeslider" className="rangeslider-label">{"\n"}
                {label}{"\n"}
                {tooltip &&
                <>
                    <i className="material-icons" data-tooltip="Some informational text" data-tooltip-position="top">{"\n"}
                        help_outline{"\n"}
                    </i>{"\n"}
                </>}
            </label>{"\n"}
            <input {...attrs}/>{"\n"}

            <ValueLabel />
            <div className="value-range">{"\n"}
                <small>{min}{valueLabelPostfix ? valueLabelPostfix : null}</small>{"\n"}
                <small>{max}{valueLabelPostfix ? valueLabelPostfix : null}</small>{"\n"}
            </div>
        </div>
    );
};

Rangeslider.propTypes = {
    name: PropTypes.string,
    label: PropTypes.string,
    min: PropTypes.number,
    max: PropTypes.number,
    step: PropTypes.number,
    value: PropTypes.number,
    disabled: PropTypes.bool,
    valueLabelPrefix: PropTypes.string,
    valueLabelPostfix: PropTypes.string
};

export default Rangeslider;
