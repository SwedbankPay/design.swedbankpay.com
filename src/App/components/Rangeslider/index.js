import React from "react";
import PropTypes from "prop-types";

const Rangeslider = ({ label, name, min, max, step, value, disabled, valueLabelPrefix, valueLabelPostfix, tooltip }) => {
    const attrs = {
        type: "range",
        name,
        min,
        max,
        step,
        defaultValue: value,
        disabled
    };

    return (
        <div className={`rangeslider ${disabled ? "disabled" : ""}`}>{"\n"}
            <input {...attrs}/>{"\n"}
            <div className="d-flex justify-content-between">{"\n"}
                <label htmlFor="rangeslider" className="rangeslider-label">{"\n"}
                    {label}{"\n"}
                    {tooltip &&
                        <>
                            <i className="material-icons" data-tooltip="Some informational text" data-tooltip-position="top">{"\n"}
                                help_outline{"\n"}
                            </i>{"\n"}
                        </>}
                </label>{"\n"}
                <output className="value-label" >{"\n\t\t"}
                    <p>{"\n\t\t\t"}
                        {valueLabelPrefix ? <span>{valueLabelPrefix}</span> : null}{valueLabelPrefix ? "\n\t\t\t" : null}
                        <span data-rs-value>{value}</span>{"\n\t\t\t"}
                        {valueLabelPostfix ? <span>{valueLabelPostfix}</span> : null}{valueLabelPostfix ? "\n\t\t" : null}
                    </p>
                </output>
            </div>
            <div className="value-range">{"\n"}
                <small>{min}{valueLabelPostfix ? valueLabelPostfix : null}</small>{"\n"}
                <small>{max}{valueLabelPostfix ? valueLabelPostfix : null}</small>{"\n"}
            </div>
        </div>
    );
};

Rangeslider.propTypes = {
    id: PropTypes.string,
    name: PropTypes.string,
    label: PropTypes.string,
    min: PropTypes.number,
    max: PropTypes.number,
    step: PropTypes.number,
    value: PropTypes.number,
    disabled: PropTypes.bool,
    valueLabelPrefix: PropTypes.string,
    valueLabelPostfix: PropTypes.string
};

export default Rangeslider;
