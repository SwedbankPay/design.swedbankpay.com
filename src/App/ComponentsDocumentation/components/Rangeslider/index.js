import React, { useEffect } from "react";
import { Link } from "react-router-dom";

import { ComponentPreview, DocContainer } from "@docutils";
import { howToUse, sliderShowCase } from "./constants";
import Tag from "@components/Tag";
import RangesliderComponent from "~/src/App/components/Rangeslider";

import { rangeslider } from "@src/scripts/main";
import NpmInformation from "@docutils/NpmInformation";

const Overview = () => (
    <section>
        <h2 id="overview">Overview</h2>
        <ComponentPreview language="html" showCasePanel showCasePanelAdvanced={sliderShowCase} codeFigure />

        <h3>When to consider something else</h3>
        <ul className="list list-bullet">
            <li>Use a <Link to="/components/input-field">text input</Link> instead when the user needs to select specific number.</li>
        </ul>

        <h3>How to use sliders</h3>

        {howToUse.map(({ type, component, text }, i) => (
            <React.Fragment key={i}>
                <div className={`slab slab-plain slab-border-${type} h-100 mb-2`}>
                    <h3 className="mt-0">Do{type === "error" && "nâ€™t"}</h3>
                    <div className="d-flex flex-column align-items-center">
                        <div className="py-3 col-12 col-lg-8">
                            {component}
                        </div>
                    </div>
                </div>
                {text}
            </React.Fragment>
        ))}
        <DisableState/>
    </section>
);

const ContentGuidelines = () => (
    <section>
        <h2 id="content-guidelines">Content guidelines</h2>

        <div className="slab slab-plain h-100 mb-2">
            <div className="d-flex flex-column align-items-center">
                <div className="py-3 col-12 col-lg-8">
                    <div>
                        <div className="rangeslider">
                            <input type="range"/>
                            <div className="d-flex justify-content-between">
                                <p className="rangeslider-label">Label</p>
                                <output className="value-label px-2 align-items-center">
                                    Selected Value
                                </output>
                            </div>
                            <div className="value-range">
                                <small>Minimum Value</small>
                                <small>Maximum Value</small>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        </div>

        <ul className="list list-bullet">
            <li>The <b>label</b> should clearly explain the context of the slider</li>
            <li>Make sure to align the <b>values</b> so that they all follow the same structure and not mixed values or symbols</li>
        </ul>
    </section>
);

const DeveloperDocumentation = () => (
    <section>
        <h2 id="developer-documentation mb-3">Developer documentation</h2>
        <h3>Javascript methods</h3>
        <p className="my-3">Use <Tag code type="secondary" text="dg.rangeslider.init()"/> to <a href="/get-started/for-developers" >initialize</a> all rangesliders.</p>
        <p>Or <Tag code type="secondary" text='dg.rangeslider.init("component-id")'/> to initialize one specific component.</p>
    </section>
);

const DisableState = () => (
    <>
        <h3>Disabled state</h3>
        <p>Disable a slider by adding <Tag code type="primary" text="disabled"/> attribute to the desired slider. </p>
        <ComponentPreview language="html" showCasePanel codeFigure>
            <div className = "d-flex flex-column align-items-center">
                <div className="py-3 col-12 col-lg-8">
                    <RangesliderComponent id="disabled_state_example" label="Label" min={0} max={100} value={50} disabled/>
                </div>
            </div>
        </ComponentPreview>
    </>
);

const Rangeslider = () => {
    useEffect(() => {
        rangeslider.init();
    });

    return (
        <DocContainer>
            <p className="lead">The slider component lets the user increase and decrease the value by moving the slider along a horizontal line and lets the user choose a value</p>
            <Overview />
            <ContentGuidelines />
            <DeveloperDocumentation />
            <NpmInformation componentName={"rangeslider"}/>
        </DocContainer>
    );
};

export default Rangeslider;

export { Overview, ContentGuidelines, DeveloperDocumentation };
