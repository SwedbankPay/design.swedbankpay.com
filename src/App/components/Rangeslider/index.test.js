import React from "react";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

import Rangeslider from ".";

describe("Component: Rangeslider -", () => {
	it("is defined", () => {
		expect(Rangeslider).toBeDefined();
	});

	it("renders with a value label", () => {
		render(<Rangeslider />);

		expect(screen.getByRole("slider")).toBeInTheDocument();
		expect(screen.getByRole("status")).toHaveClass("value-label");
		expect(
			screen.getByRole("status").querySelectorAll("[data-rs-value]")
		).toHaveLength(1);

		const componentForSnap = renderer.create(<Rangeslider />);

		expect(componentForSnap.toJSON()).toMatchSnapshot();
	});

	it("renders with a value label and a value label prefix", () => {
		render(<Rangeslider valueLabelPrefix="test" />);

		expect(screen.getByRole("slider")).toBeInTheDocument();
		expect(screen.getByRole("status").querySelector("span")).toHaveTextContent(
			"test"
		);
		expect(screen.getByRole("status")).toHaveClass("value-label");

		const componentForSnap = renderer.create(
			<Rangeslider valueLabelPrefix="test" />
		);

		expect(componentForSnap.toJSON()).toMatchSnapshot();
	});

	it("renders with a value label and a value label postfix", () => {
		render(<Rangeslider valueLabelPostfix="test" />);

		expect(screen.getByRole("slider")).toBeInTheDocument();
		expect(
			[...screen.getByRole("status").querySelectorAll("span")].filter(
				(elmt) => !elmt.dataset.rsValue
			)[0]
		).toHaveTextContent("test");
		expect(screen.getByRole("status")).toHaveClass("value-label");

		const componentForSnap = renderer.create(
			<Rangeslider valueLabelPostfix="test" />
		);

		expect(componentForSnap.toJSON()).toMatchSnapshot();
	});
});
