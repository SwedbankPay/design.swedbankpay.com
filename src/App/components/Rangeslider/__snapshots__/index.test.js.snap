// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Rangeslider - renders with a value label 1`] = `
<Rangeslider>
  <div
    className="rangeslider"
  >
    

    <input
      type="range"
    />
    

    <div>
      

      <label
        className="rangeslider-label"
      >
        

        

      </label>
      

      <output
        className="value-label"
      >
        
		
        <span
          data-rs-value={true}
        />
        
			
      </output>
    </div>
    <div
      className="value-range"
    >
      

      <small />
      

      <small />
      

    </div>
  </div>
</Rangeslider>
`;

exports[`Component: Rangeslider - renders with a value label and a value label postfix 1`] = `
<Rangeslider
  valueLabelPostfix="test"
>
  <div
    className="rangeslider"
  >
    

    <input
      type="range"
    />
    

    <div>
      

      <label
        className="rangeslider-label"
      >
        

        

      </label>
      

      <output
        className="value-label"
      >
        
		
        <span
          data-rs-value={true}
        />
        
			
        <span>
          test
        </span>
        
		
      </output>
    </div>
    <div
      className="value-range"
    >
      

      <small>
        test
      </small>
      

      <small>
        test
      </small>
      

    </div>
  </div>
</Rangeslider>
`;

exports[`Component: Rangeslider - renders with a value label and a value label prefix 1`] = `
<Rangeslider
  valueLabelPrefix="test"
>
  <div
    className="rangeslider"
  >
    

    <input
      type="range"
    />
    

    <div>
      

      <label
        className="rangeslider-label"
      >
        

        

      </label>
      

      <output
        className="value-label"
      >
        
		
        <span>
          test
        </span>
        
			
        <span
          data-rs-value={true}
        />
        
			
      </output>
    </div>
    <div
      className="value-range"
    >
      

      <small>
        test
      </small>
      

      <small>
        test
      </small>
      

    </div>
  </div>
</Rangeslider>
`;
